### PR TITLE
[SPARK-41677][CORE][SQL][SS][UI] Add Protobuf serializer for `StreamingQueryProgressWrapper`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
Add Protobuf serializer for `StreamingQueryProgressWrapper `

### Why are the changes needed?
Support fast and compact serialization/deserialization for `StreamingQueryProgressWrapper` over RocksDB.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Add new UT
- Manual test sql module with `LIVE_UI_LOCAL_STORE_DIR`, all test passed: 

```
build/mvn clean install -DskipTests -pl sq/core -am
export LIVE_UI_LOCAL_STORE_DIR=/tmp/spark-ui
build/mvn clean install -pl sql/core
```

There 4 existing test suites classes use `StreamingQueryProgressWrapper`:
 - StateStoreCoordinatorSuite
 - StreamingQueryStatusListenerWithDiskStoreSuite
 - UISeleniumSuite
 - UISeleniumWithRocksDBBackendSuite

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 1]
```diff
@@ -125,14 +127,16 @@ class StateOperatorProgress private[sql](
...
   val id: UUID,
   val runId: UUID,
   val name: String,
   val timestamp: String,
   val batchId: Long,
   val batchDuration: Long,
   val durationMs: ju.Map[String, JLong],
+  // If `eventTime` is no longer read-only, need change
+  // `StreamingQueryProgressSerializer#deserialize` to return mutable Java Map
   val eventTime: ju.Map[String, String],
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala:140]: There are 6 `ju.Map`,  `durationMs` and `observedMetrics` already use mutable Java Map, the remaining 4 Java Map should be read-only at present, but for safety, we can also directly return mutable Java Map in the deserialize method

**Comment by LuciferYang** [comment]
Will refactor after https://github.com/apache/spark/pull/39666 merged

**Comment by LuciferYang** [comment]
rebased

**Comment by LuciferYang** [comment]
cc @gengliangwang 

Does this one have a chance to Spark 3.4.0 ? Or wait for the next version?

File: core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto [hunk 2]
```diff
@@ -765,3 +765,54 @@ message PoolData {
...
+  int64 num_rows_updated = 3;
+  int64 all_updates_time_ms = 4;
+  int64 num_rows_removed = 5;
+  int64 all_removals_time_ms = 6;
+  int64 commit_time_ms = 7;
+  int64 memory_used_bytes = 8;
+  int64 num_rows_dropped_by_watermark = 9;
+  int64 num_shuffle_partitions = 10;
+  int64 num_state_store_instances = 11;
+  map<string, int64> custom_metrics = 12;
```
Comment (gengliangwang) [inline core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto:781]: I am concerned about the nullability of all these maps. Shall we check/test all of them?

File: sql/core/src/test/scala/org/apache/spark/status/protobuf/sql/KVStoreProtobufSerializerSuite.scala [hunk 3]
```diff
@@ -271,4 +278,254 @@ class KVStoreProtobufSerializerSuite extends SparkFunSuite {
       assert(result.endTimestamp == input.endTimestamp)
     }
   }
+
+  test("StreamingQueryProgressWrapper") {
+    val normalInput = {
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/status/protobuf/sql/KVStoreProtobufSerializerSuite.scala:283]: Two objects are manually created due to many fields can be null

File: sql/core/src/test/scala/org/apache/spark/status/protobuf/sql/KVStoreProtobufSerializerSuite.scala [hunk 4]
```diff
@@ -271,4 +278,254 @@ class KVStoreProtobufSerializerSuite extends SparkFunSuite {
...
+      val progress = input.progress
+      val resultProcess = result.progress
+      assert(progress.id == resultProcess.id)
+      assert(progress.runId == resultProcess.runId)
+      assert(progress.name == resultProcess.name)
+      assert(progress.timestamp == resultProcess.timestamp)
+      assert(progress.batchId == resultProcess.batchId)
+      assert(progress.batchDuration == resultProcess.batchDuration)
+      if (hasNullValue) {
+        assert(resultProcess.durationMs.isEmpty)
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/status/protobuf/sql/KVStoreProtobufSerializerSuite.scala:465]: null map input return empty map

File: core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto (hunk 2 above)
Comment (LuciferYang) [reply core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto:781]: [a904a27](https://github.com/apache/spark/pull/39642/commits/a904a27919a47cebf3784a8756f46b3237b4be46) check/test all map 

[699ebd1](https://github.com/apache/spark/pull/39642/commits/699ebd1e6c3905722d0b09ff11e5dccc31813d3c) add  `setJMapField`  function to `Utils`

File: sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+private[protobuf] object StreamingQueryProgressSerializer {
+
+  private val mapper: JsonMapper = JsonMapper.builder()
+    .addModule(DefaultScalaModule)
+    .build()
+
+  def serialize(process: StreamingQueryProgress): StoreTypes.StreamingQueryProgress = {
+    import org.apache.spark.status.protobuf.Utils.setJMapField
+    val builder = StoreTypes.StreamingQueryProgress.newBuilder()
+    if (process.id != null) {
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala:40]: use setStringField?

File: sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      builder.setId(process.id.toString)
+    }
+    if (process.runId != null) {
+      builder.setRunId(process.runId.toString)
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala:44]: ditto

File: sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    builder.setBatchDuration(process.batchDuration)
+    setJMapField(process.durationMs, builder.putAllDurationMs)
+    setJMapField(process.eventTime, builder.putAllEventTime)
+    process.stateOperators.foreach(
+      s => builder.addStateOperators(StateOperatorProgressSerializer.serialize(s)))
+    process.sources.foreach(
+      s => builder.addSources(SourceProgressSerializer.serialize(s))
+    )
+    builder.setSink(SinkProgressSerializer.serialize(process.sink))
+    if (process.observedMetrics != null && !process.observedMetrics.isEmpty) {
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala:58]: use setJMapField?

File: sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala (hunk 5 above)
Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala:40]: `process.id` is UUID, cannot be the first parameter of `setStringField `

File: sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala (hunk 6 above)
Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala:44]: ditto

File: sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    builder.setBatchDuration(process.batchDuration)
+    setJMapField(process.durationMs, builder.putAllDurationMs)
+    setJMapField(process.eventTime, builder.putAllEventTime)
+    process.stateOperators.foreach(
+      s => builder.addStateOperators(StateOperatorProgressSerializer.serialize(s)))
+    process.sources.foreach(
+      s => builder.addSources(SourceProgressSerializer.serialize(s))
+    )
+    builder.setSink(SinkProgressSerializer.serialize(process.sink))
+    def serializeMetrics(metrics: JMap[String, Row]): Unit = {
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala:57]: use `setJMapField` need to define an additional function or use an anonymous function due this is not a simple `putAll` scenario.

Any other better way to write this ? @gengliangwang

File: sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala (hunk 5 above)
Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala:39]: https://github.com/apache/spark/blob/c43be4eeeea1b23b599ef9c7f81c2cba619bcf7e/core/src/main/scala/org/apache/spark/status/protobuf/Utils.scala#L27-L31

Or should we refactor `setStringField` to 

```scala
  def setStringField[T](input: T, f: String => Any): Unit = {
    if (input != null) {
      f(input.toString)
    }
  }
```

File: sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+      eventTime = new JHashMap(process.getEventTimeMap),
+      stateOperators =
+        StateOperatorProgressSerializer.deserializeToArray(process.getStateOperatorsList),
+      sources = SourceProgressSerializer.deserializeToArray(process.getSourcesList),
+      sink = SinkProgressSerializer.deserialize(process.getSink),
+      observedMetrics = convertToObservedMetrics(process.getObservedMetricsMap)
+    )
+  }
+
+  private def putAllObservedMetrics(
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala:85]: Change to define a class level private function

File: sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala (hunk 5 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala:39]: nvm, let's keep the current code

File: sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      builder: StoreTypes.StreamingQueryProgress.Builder,
+      observedMetrics: JMap[String, Row]): Unit = {
+    observedMetrics.forEach {
+      case (k, v) => builder.putObservedMetrics(k, mapper.writeValueAsString(v))
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/status/protobuf/sql/StreamingQueryProgressSerializer.scala:89]: nit: Let's add one line comment to mention why we choose to encode the row object with json.

**Review by gengliangwang — APPROVED** [review]
LGTM. Thanks for working on this during holiday!

**Comment by LuciferYang** [comment]
done. Thank you for reviewing the code during the holiday :) @gengliangwang

**Comment by LuciferYang** [comment]
GA passed

**Comment by LuciferYang** [comment]
Thanks @gengliangwang

**Pull request closed without merging**